File: kmp-workshop/iosApp/iosApp/BookListView.swift
//
// Created by Dominik Chmelík on 14.10.2025.
//

import SwiftUI

struct BookListView: View {
    @State private var viewModel: BookViewModel = BookViewModel()
    @State private var presentedBook: Book?

    var body: some View {
        List(viewModel.books, id: \.id) { book in
            Text(book.name)
                .fontWeight(.heavy)
                .onTapGesture {
                    presentedBook = book
                }
        }
        .task {
            viewModel.fetchBooks()
        }
        .sheet(item: $presentedBook) { book in
            BookDetailView(viewModel: $viewModel, book: book)
        }
    }
}

File: cmp-workshop/composeApp/src/androidMain/kotlin/com/chmelik/cmp_workshop/BookListView.kt
package com.chmelik.cmp_workshop

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.SheetState
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BookListView(
    viewModel: BookViewModel
) {
    var presentedBook by rememberSaveable { mutableStateOf<Book?>(null) }
    val sheetState: SheetState = rememberModalBottomSheetState()

    LaunchedEffect(Unit) {
        viewModel.fetchBooks()
    }

    LazyColumn {
        items(items = viewModel.books, key = { it.id }) { book ->
            Text(
                text = book.name,
                fontWeight = FontWeight.Black,
                modifier = Modifier
                    .clickable { presentedBook = book }
                    .padding(16.dp)
            )
        }
    }

    presentedBook?.let {
        ModalBottomSheet(
            onDismissRequest = { presentedBook = null },
            sheetState = sheetState
        ) {
            BookDetailView(
                book = it,
                viewModel = viewModel
            )
        }
    }
}
